### PR TITLE
Add vendor package files for orocos_kdl

### PR DESCRIPTION
**Pull request by jacobperron**
Build orocos_kdl as an external project if it is not available on the system.

Related to https://github.com/ros2/ros2/pull/1208

### CODE REVIEW

**Review by nuclearsandwich — APPROVED** [review]
One small question but otherwise looks like a standard vendor package.

File: CMakeLists.txt
Comment (jacobperron) [reply CMakeLists.txt:13]: It's lower on my priority list, since we are not currently depending on on the python bindings. If I don't get to it before merging this PR, then I'll create a ticket.

Comment (clalancette) [reply CMakeLists.txt:13]: There's a lot of users who actually want the Python bindings here.  Porting of the python bindings for e.g. tf2_kdl has been blocked on this for a long time.

That said, I'm fine with starting without that functionality, and then enabling it in a follow-up pull request.

Comment (jacobperron) [reply CMakeLists.txt]: Okay, I spent some time looking at vendoring the Python bindings.

I've broken it up into two vendor packages, `orocos_kdl_vendor` and `python_orocos_kdl_vendor`. The latter depends on the former. This was more straight-forward than trying to link one CMake external project on another and it's more flexible since users who don't care about Python support don't have to build/install it.

In order to get the recent upstream changes to skip building pybind11, I'm now pointing to a newer commit (dbee885).

I'm able to build both packages locally (if system packages for orocod_kdl are not installed), however I'm still not able to import PyKDL. I'm not sure if it's something that should be fixed in the vendor package or upstream; further investigation required.

Comment (jacobperron) [reply CMakeLists.txt]: Okay, sothe problem is that the shared library for the Python bindings is not a part of PYTHONPATH after sourcing the workspace. It seems pykdl relies on catkin for the env hook (see https://github.com/orocos/orocos_kinematics_dynamics/pull/285). I can add an env hook to the vendor package, but I'm not sure how to determine the path of the library in the binary directory; [it is dependent on the Python version](https://github.com/orocos/orocos_kinematics_dynamics/blob/4f705fe9e1533802578b5275ddad02b065b2be4e/python_orocos_kdl/CMakeLists.txt#L28). There must be some cmake spell to get this done, but I guess we could also duplicate the logic from upstream, though it seems brittle. If `catkin` is installed, then the install path changes.

Comment (jacobperron) [reply CMakeLists.txt]: I've determined how to get the location of the shared library in the binary directory easily, with `ExternalProject_Get_Property`. Next, I tried to install it with [ament_python_install_module](https://github.com/ament/ament_cmake/blob/199ac7a046cbf686912a198a41a501858382af60/ament_cmake_python/cmake/ament_python_install_module.cmake), which _almost_ works except that the file doesn't exist at the configure step.. so the build never happens. Here's what the Cmake looks like:

```cmake
  # Build PyKDL
  externalproject_add(pykdl
    URL https://github.com/orocos/orocos_kinematics_dynamics/archive/e25a13fc5820dbca6b23d10506407bca9bcdd25f.zip
    URL_MD5 8ac40c9f84eca10dbe1f3777ba715ba0
    TIMEOUT 600 
    SOURCE_SUBDIR python_orocos_kdl
    INSTALL_COMMAND ""
    CMAKE_ARGS
      ${extra_cmake_args}
  )

  # Install Python module
  ExternalProject_Get_Property(pykdl BINARY_DIR)
  if(WIN32)
    set(pykdl_module "${BINARY_DIR}/PyKDL.pyd")
  else()
    set(pykdl_module "${BINARY_DIR}/PyKDL.so")
  endif()

  ament_python_install_module(${pykdl_module}
```

:confused:

Comment (jacobperron) [reply CMakeLists.txt]: Okay, I just copied the relevant parts from `ament_python_install_module` since we don't seem to be able to call it directly.

@clalancette @nuclearsandwich PTAL [556def6](https://github.com/ros2/orocos_kdl_vendor/pull/1/commits/556def6bdcb8b008fe45a22c1ed5ebb4c964b3ad)

Comment (nuclearsandwich) [reply CMakeLists.txt]: > I can add an env hook to the vendor package, but I'm not sure how to determine the path of the library in the binary directory; [it is dependent on the Python version](https://github.com/orocos/orocos_kinematics_dynamics/blob/4f705fe9e1533802578b5275ddad02b065b2be4e/python_orocos_kdl/CMakeLists.txt#L28). There must be some cmake spell to get this done, but I guess we could also duplicate the logic from upstream, though it seems brittle. If `catkin` is installed, then the install path changes.

One solution is to add optional ament_cmake handling to upstream directly so that it could manage its own env hooks on ROS 2 as well. I think it could co-exist with the catkin setup in package.xml through the use of conditional dependencies. If we did that, it would make it possible to release upstream directly into ROS 2, which may obviate the need for vendor packages entirely at the cost of removing our ability to work with the platform version of orocos_kdl.

Comment (jacobperron) [reply CMakeLists.txt]: I'm okay with that option, though is there a concern about side-by-side installing the version of orocos_kdl shipped with ROS the upstream version?

Comment (jacobperron) [reply CMakeLists.txt]: I noticed that orocos_kdl is not released in Noetic (I guess maybe we're depending on upstream in that case?).

@MatthijsBurgh Maybe maintainers of orocos_kdl have a preference? Would it be better for ROS 2 users to depend on the version of orocos_kdl shipped with Ubuntu or a version released into ROS 2 itself? The latter option requires someone make regular releases into ROS 2.

Comment (nuclearsandwich) [reply CMakeLists.txt]: > If we did that, it would make it possible to release upstream directly into ROS 2, which may obviate the need for vendor packages entirely at the cost of removing our ability to work with the platform version of orocos_kdl.

This contradicts the conclusions (and my own agreement with them) in https://github.com/ros2/ros2/pull/1208#issuecomment-1007533586

I'm trying to unpack my own attitude. I think it comes down to how much logic we want in the vendor package itself. The vendor package in its ideal form is just a wrapper around an upstream project which provides the glue to build it on platforms which don't currently provide it.

The conflict for me is between logicless vendor packages (just enough to get the vendored package built and bolted in) and keeping packages _out_ of ROS when we could get them as platform-provided dependencies.

If in the long view we're trying to keep orocos_kdl as a platform dependency and willing, as maintainers and as a community, to deal with the trade-offs of working with platforms development cycles then we should try to make the vendor packages work since hopefully someday they'll be gone when all our tier 1 & 2 platforms provide them.
If we're expecting to want to develop and maintain orocos_kdl together with ROS 2 the way we do with some rmw providers then the equation flips and we may as well embrace it as a native ROS package now in order to avoid dealing with this vendor logic. My current understanding is that we'd be very happy with orocos_kdl from upstream and thus we should try to figure out how to keep the vendor logic low, but also not give up on the vendor packages.

Comment (sloretz) [reply CMakeLists.txt]: > I'm not sure how to determine the path of the library in the binary directory; [it is dependent on the Python version](https://github.com/orocos/orocos_kinematics_dynamics/blob/4f705fe9e1533802578b5275ddad02b065b2be4e/python_orocos_kdl/CMakeLists.txt#L28)

Opened https://github.com/ros2/orocos_kdl_vendor/pull/2 to solve that problem using `FetchContent`, `add_subdirectory()`, and `get_directory_property()`.

Comment (MatthijsBurgh) [reply CMakeLists.txt]: I recommend to only use 1.5.X versions. Ubuntu has only released a 1.5.X on Jammy (22.04). So I think it would be best to release one by ROS2.

**Comment by jacobperron** [comment]
I've been neglecting this ticket, but last I checked it was building successfully. Now I see the following CMake error from `python_orocos_kdl_vendor` :confused: 

```
CMake Error at /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:146 (message):
  Could NOT find PythonLibs (missing: PYTHON_INCLUDE_DIRS) (Required is at
  least version "3.8")
Call Stack (most recent call first):
  /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:393 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake-3.16/Modules/FindPythonLibs.cmake:310 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  /home/jacob/ws/ros/orocos_upstream/build/python_orocos_kdl_vendor/_deps/python_orocos_kdl-src/python_orocos_kdl/CMakeLists.txt:25 (find_package)
```

I don't think anything code changed...

@sloretz Did this work for you previously on Linux if the upstream package is not installed?

**Comment by sloretz** [comment]
>   Could NOT find PythonLibs (missing: PYTHON_INCLUDE_DIRS) (Required is at
  least version "3.8")

Is this a fresh container? Is either `python3-dev` or `libpython3-dev` installed?

> @sloretz Did this work for you previously on Linux if the upstream package is not installed?

I forget how I tested it, but I can try again in a bit.

**Comment by jacobperron** [comment]
I'm not in a container (native 20.04), so it's possible some systems dependencies were updated. Both `python3-dev` and `libpython3-dev` are installed. I can also build the package without the vendor package in-the-loop (cloning [orocos_kdl](https://github.com/orocos/orocos_kinematics_dynamics) into it's own workspace and running `colcon build`) :shrug:

**Comment by jacobperron** [comment]
Okay, it looks like the `find_package(pybind11)` is causing some interference. Here's a CMakeLists.txt that reproduces the error for me:

```
cmake_minimum_required(VERSION 3.5)
project(foo)
find_package(pybind11)
find_package(PythonInterp 3 REQUIRED)
find_package(PythonLibs 3.8 REQUIRED)
```

**Comment by jacobperron** [comment]
[976cc59](https://github.com/ros2/orocos_kdl_vendor/pull/1/commits/976cc59c32cd5678624c3052fc480efced9ee8d6) fixes things for me.

**Comment by sloretz** [comment]
> I'm not in a container (native 20.04), so it's possible some systems dependencies were updated. Both python3-dev and libpython3-dev are installed. I can also build the package without the vendor package in-the-loop (cloning [orocos_kdl](https://github.com/orocos/orocos_kinematics_dynamics) into it's own workspace and running colcon build) shrug

Interesting. I can't reproduce.

I did a clean build in two containers: a Jammy container I created yesterday morning, and a Focal container that I created a long time ago with freshly `apt upgrade`d packages today. Both builds were successful.

`colcon build --packages-above-and-dependencies orocos_kdl_vendor` 

Here are the branches I used

<details><summary>ros2.repos diff</summary>

Mostly - I have a few extra repos in my workspace but I didn't include that part of the diff.

```diff
@@ -170,7 +190,7 @@ repositories:
   ros/kdl_parser:
     type: git
     url: https://github.com/ros/kdl_parser.git
-    version: ros2
+    version: jacob/upstream_orocos_kdl
   ros/pluginlib:
     type: git
     url: https://github.com/ros/pluginlib.git
@@ -182,7 +202,7 @@ repositories:
   ros/robot_state_publisher:
     type: git
     url: https://github.com/ros/robot_state_publisher.git
-    version: ros2
+    version: jacob/upstream_orocos_kdl
   ros/ros_environment:
     type: git
     url: https://github.com/ros/ros_environment.git
@@ -230,7 +250,7 @@ repositories:
   ros2/geometry2:
     type: git
     url: https://github.com/ros2/geometry2.git
-    version: ros2
+    version: jacob/upstream_orocos_kdl
   ros2/launch:
     type: git
     url: https://github.com/ros2/launch.git
@@ -251,6 +271,10 @@ repositories:
     type: git
     url: https://github.com/ros2/mimick_vendor.git
     version: master
+  ros2/orocos_kdl_vendor:
+    type: git
+    url: git@github.com:ros2/orocos_kdl_vendor.git
+    version: jacob/package_files
```
</details>

> Okay, it looks like the find_package(pybind11) is causing some interference. Here's a CMakeLists.txt that reproduces the error for me:

I'll give this a try next

**Review by sloretz — APPROVED** [review]
`colcon build --packages-up-to python_orocos_kdl_vendor` on Jammy and Focal succeeds for me with and without https://github.com/ros2/orocos_kdl_vendor/pull/1/commits/976cc59c32cd5678624c3052fc480efced9ee8d6 :shrug:

**Comment by sloretz** [comment]
>     Okay, it looks like the find_package(pybind11) is causing some interference. Here's a CMakeLists.txt that reproduces the error for me:
> 
> I'll give this a try next

Interesting, with this smaller package I do see the same error on both Jammy and Focal. I had to add a little hack to trick colcon into building `pybind11_vendor` for `foo` to be able to find_package it.

```CMakeList.txt
cmake_minimum_required(VERSION 3.5)
project(foo)

set(dummy "find_package(pybind11_vendor)")
find_package(pybind11)
find_package(PythonInterp 3 REQUIRED)
find_package(PythonLibs 3.8 REQUIRED)
```

**Comment by jacobperron** [comment]
Do you have `pybind11-dev` installed? I do:
```$ apt-cache policy pybind11-dev
pybind11-dev:
  Installed: 2.4.3-2build2
  Candidate: 2.4.3-2build2
```

**Comment by sloretz** [comment]
> Do you have pybind11-dev installed? I do:

I have it installed on the Jammy container

```console
$ apt list --installed | grep pybind11

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

pybind11-dev/jammy,now 2.9.1-2 all [installed]
```

But not the focal container

**Comment by sloretz** [comment]
I don't see the `CMP0054` warning anymore either, which was the point of the `find_package(pybind11)` call in the first place, so 976cc59 looks fine to me!

File: python_orocos_kdl_vendor/CMakeLists.txt
Comment (clalancette) [inline python_orocos_kdl_vendor/CMakeLists.txt:44]: This can be completely ignored, but I do think CMake 3.14 can be made the minimum here.  This package will only ever be used for Humble and forwards, and looking at https://www.ros.org/reps/rep-2000.html#humble-hawksbill-may-2022-may-2027 , all of the platforms have CMake >= 3.14.

File: python_orocos_kdl_vendor/package.xml
Comment (clalancette) [inline python_orocos_kdl_vendor/package.xml:10]: This isn't strictly true anymore, since we are using `FetchContent` rather than `ExternalProject`.  But it is nitpicking.

**Review by clalancette — APPROVED** [review]
I've left a few nits inline, but this otherwise looks good to me with green CI.

File: python_orocos_kdl_vendor/CMakeLists.txt
Comment (sloretz) [reply python_orocos_kdl_vendor/CMakeLists.txt:44]: I wasn't sure about the `3.16.1 / 3.12.2` mention in the OpenEmbedded column. Does that mean there's a version of OpenEmbedded where the CMake version is 3.12.2?

File: python_orocos_kdl_vendor/package.xml
Comment (jacobperron) [reply python_orocos_kdl_vendor/package.xml]: Updated [e21ad5b](https://github.com/ros2/orocos_kdl_vendor/pull/1/commits/e21ad5b2125ae8abcd7a0a96bda9d5ef90901cab)[](https://github.com/jacobperron)

File: python_orocos_kdl_vendor/CMakeLists.txt
Comment (jacobperron) [reply python_orocos_kdl_vendor/CMakeLists.txt:44]: I guess we'll hold off on changing this unless REP-2000 is updated.

Comment (clalancette) [reply python_orocos_kdl_vendor/CMakeLists.txt:44]: Yeah, that's fine.  It's a really minor point, I'm fine with the workaround we have in place.

**Review by sloretz — APPROVED** [review]

**Comment by jacobperron** [comment]
The CMP0054 warning was fixed upstream (thanks @MatthijsBurgh for the quick review and merge!). I think this one is good to go now.

Here's CI for all connected PRs:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/16517/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11104/)
* Windows [Build Status](https://ci.ros2.org/job/ci_windows/16893/)

**Review by sloretz — APPROVED** [review]

**Pull request merged**
